In [19]:
import yfinance as yf
import pandas as pd
import plotly.express as px

In [20]:
aapl= yf.download("AAPL", period="1y",auto_adjust=True)
aapl.head()

[*********************100%***********************]  1 of 1 completed


Price,Close,High,Low,Open,Volume
Ticker,AAPL,AAPL,AAPL,AAPL,AAPL
Date,,,,,
2025-07-15,208.283691,211.052705,208.094440,208.393257,42296300
2025-07-16,209.329544,211.560683,207.815546,209.468990,47490500
2025-07-17,209.190094,210.963059,208.761785,209.737924,48068100
2025-07-18,210.345505,210.953095,208.871357,210.036732,48974600
2025-07-21,211.640381,214.927344,210.793749,211.261893,51377400


In [21]:
tsla = yf.download("TSLA", period="1y", auto_adjust=True)
tsla.tail()

[*********************100%***********************]  1 of 1 completed


Price,Close,High,Low,Open,Volume
Ticker,TSLA,TSLA,TSLA,TSLA,TSLA
Date,,,,,
2026-07-08,394.059998,399.630005,390.510010,399.380005,33844900
2026-07-09,406.549988,407.859985,390.859985,393.989990,37835000
2026-07-10,407.760010,413.160004,402.809998,410.489990,33410000
2026-07-13,394.760010,405.570007,391.369995,404.609985,32809200
2026-07-14,NaN,NaN,NaN,NaN,23374603


In [22]:
print(aapl.columns)

MultiIndex([( 'Close', 'AAPL'),
            (  'High', 'AAPL'),
            (   'Low', 'AAPL'),
            (  'Open', 'AAPL'),
            ('Volume', 'AAPL')],
           names=['Price', 'Ticker'])


In [23]:
aapl.columns = aapl.columns.get_level_values(0)
tsla.columns = tsla.columns.get_level_values(0)
aapl.head()

Price,Close,High,Low,Open,Volume
Date,,,,,
2025-07-15,208.283691,211.052705,208.094440,208.393257,42296300
2025-07-16,209.329544,211.560683,207.815546,209.468990,47490500
2025-07-17,209.190094,210.963059,208.761785,209.737924,48068100
2025-07-18,210.345505,210.953095,208.871357,210.036732,48974600
2025-07-21,211.640381,214.927344,210.793749,211.261893,51377400


In [24]:
print(aapl.columns)

Index(['Close', 'High', 'Low', 'Open', 'Volume'], dtype='str', name='Price')


In [25]:
print("Columns:", list(aapl.columns))
print("Index type:", type(aapl.index))
print("AAPL rows:", len(aapl))
print("TSLA rows:", len(tsla))
print("Date range:", aapl.index.min(), "→", aapl.index.max())

Columns: ['Close', 'High', 'Low', 'Open', 'Volume']
Index type: <class 'pandas.DatetimeIndex'>
AAPL rows: 251
TSLA rows: 251
Date range: 2025-07-15 00:00:00 → 2026-07-14 00:00:00


In [26]:
fig = px.line(aapl, y="Close", title="AAPL — Closing Price (1 Year)")
fig.show()

In [27]:
fig = px.line(tsla, y="Close", title="TSLA — Closing Price (1 Year)")
fig.show()

In [28]:
calendar_days= (aapl.index.max() - aapl.index.min()).days
print("Calendar days:", calendar_days)
print("Trading days:", len(aapl))
print("Missing days:", calendar_days - len(aapl))

Calendar days: 364
Trading days: 251
Missing days: 113


# Why are there ~252 rows instead of 365?

The dataset contains one row for each **trading day**, not every day on the calendar. Stock exchanges do not operate on **weekends** and are also closed on **official market holidays**. Because of these non-trading days, a full year typically has only **about 252 trading sessions** instead of 365 calendar days.

For example:

- Calendar days in a year: **365**
- Weekends: **~104 days**
- Market holidays: **~9–10 days**
- Remaining trading days: **≈252**

This is why a year's worth of stock data usually contains around **251–252 rows**, which is exactly what you observed.

---

# Why use adjusted closing prices?

A stock split changes the number of shares while reducing the price proportionally. For instance, in a **3-for-1 stock split**, a share priced at **$300** becomes **$100**, but shareholders receive **three times as many shares**. Their overall investment value remains unchanged.

If we use the **raw closing price**, the chart suddenly shows a massive price drop, even though no real loss occurred. A machine learning model trained on this data could mistakenly interpret the split as a significant market crash.

Using `auto_adjust=True` corrects this issue by **adjusting historical prices** so the price series remains smooth and consistent across stock splits.

The same principle applies to **dividends**. On the ex-dividend date, the stock price usually falls by approximately the dividend amount because that value is paid to shareholders. Adjusted prices account for this payment, removing the artificial drop and producing a price series that represents the stock's **actual investment performance (total return)** rather than misleading price movements.